File: reporter/src/ingress.rs
use crate::errors::ReporterError;
use crate::models::{
    ErrorResponse, SessionRequest, SessionResponse, TestcaseRequest, TestcasesRequest,
};
use reqwest::Client;

#[derive(Clone)]
pub struct IngressClient {
    client: Client,
    endpoint: String,
    api_key: String,
}

impl IngressClient {
    pub fn new(endpoint: String, api_key: String) -> Result<Self, ReporterError> {
        let client = Client::new();

        Ok(IngressClient {
            client,
            endpoint,
            api_key,
        })
    }

    pub async fn create_session(&self, session: SessionRequest) -> Result<String, ReporterError> {
        let resp = self
            .client
            .post(format!("{}/api/v1/ingress/sessions", self.endpoint))
            .header("X-API-Key", &self.api_key)
            .json(&session)
            .send()
            .await
            .map_err(|e| ReporterError::Unknown(format!("error sending session request: {}", e)))?;

        let status = resp.status();
        if !status.is_success() {
            let error_msg = match resp.text().await {
                Ok(x) => match serde_json::from_str::<ErrorResponse>(&x) {
                    Ok(err_resp) => err_resp.message,
                    Err(_) => x,
                },
                Err(_) => "".to_string(),
            };
            return Err(ReporterError::Ingress(
                format!("failed session request: {}", error_msg),
                status.as_u16(),
            ));
        }

        let session = resp.json::<SessionResponse>().await.map_err(|e| {
            ReporterError::Unknown(format!("error parsing session response: {}", e))
        })?;

        Ok(session.id)
    }

    pub async fn create_testcases(
        &self,
        testcases: Vec<TestcaseRequest>,
    ) -> Result<(), ReporterError> {
        let resp = self
            .client
            .post(format!("{}/api/v1/ingress/testcases", self.endpoint))
            .header("X-API-KEY", &self.api_key)
            .json(&TestcasesRequest { testcases })
            .send()
            .await
            .map_err(|e| {
                eprintln!("ERROR: {}", e);
                ReporterError::Unknown(format!("error sending testcase request: {}", e))
            })?;

        let status = resp.status();
        if !status.is_success() {
            let error_msg = match resp.text().await {
                Ok(x) => match serde_json::from_str::<ErrorResponse>(&x) {
                    Ok(err_resp) => err_resp.message,
                    Err(_) => x,
                },
                Err(_) => "".to_string(),
            };
            return Err(ReporterError::Ingress(
                format!("failed testcase request: {}", error_msg),
                status.as_u16(),
            ));
        }

        Ok(())
    }
}

File: reporter/ingress.go
package reporter

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

type IngressClient struct {
	client   *http.Client
	endpoint string
	apiKey   string
}

func NewIngressClient(endpoint, apiKey string) *IngressClient {
	return &IngressClient{
		client:   &http.Client{},
		endpoint: endpoint,
		apiKey:   apiKey,
	}
}

func (c *IngressClient) CreateSession(session SessionRequest) (string, error) {
	url := fmt.Sprintf("%s/api/v1/ingress/sessions", c.endpoint)

	body, err := json.Marshal(session)
	if err != nil {
		return "", NewUnknownError(fmt.Sprintf("error marshaling session request: %v", err))
	}

	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return "", NewUnknownError(fmt.Sprintf("error creating session request: %v", err))
	}

	req.Header.Set("X-API-Key", c.apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return "", NewUnknownError(fmt.Sprintf("error sending session request: %v", err))
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		bodyBytes, _ := io.ReadAll(resp.Body)
		errorMsg := string(bodyBytes)

		var errResp ErrorResponse
		if err := json.Unmarshal(bodyBytes, &errResp); err == nil {
			errorMsg = errResp.Message
		}

		return "", NewIngressError(fmt.Sprintf("failed session request: %s", errorMsg), resp.StatusCode)
	}

	var sessionResp SessionResponse
	if err := json.NewDecoder(resp.Body).Decode(&sessionResp); err != nil {
		return "", NewUnknownError(fmt.Sprintf("error parsing session response: %v", err))
	}

	return sessionResp.Id, nil
}

func (c *IngressClient) CreateTestcases(testcases []TestcaseRequest) error {
	url := fmt.Sprintf("%s/api/v1/ingress/testcases", c.endpoint)

	reqBody := TestcasesRequest{Testcases: testcases}
	body, err := json.Marshal(reqBody)
	if err != nil {
		return NewUnknownError(fmt.Sprintf("error marshaling testcase request: %v", err))
	}

	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return NewUnknownError(fmt.Sprintf("error creating testcase request: %v", err))
	}

	req.Header.Set("X-API-KEY", c.apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return NewUnknownError(fmt.Sprintf("error sending testcase request: %v", err))
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		bodyBytes, _ := io.ReadAll(resp.Body)
		errorMsg := string(bodyBytes)

		var errResp ErrorResponse
		if err := json.Unmarshal(bodyBytes, &errResp); err == nil {
			errorMsg = errResp.Message
		}

		return NewIngressError(fmt.Sprintf("failed testcase request: %s", errorMsg), resp.StatusCode)
	}

	return nil
}
